src/transpose.cpp

```diff
@@ -0,0 +1,140 @@
+#include <Rcpp.h>
+using namespace Rcpp;
+
+// copy from x with number of rows, to c column in columns
+template <typename T, class RTYPE>
+inline void copy(T column, int rows, List x, int c) {
+  for (int r = 0; r < rows; r++) {
+    List row = x[r];
+    SEXP cell = row[c];
+    if (cell != R_NilValue) {
+      column[r] = as<RTYPE>(cell);
+    }
+  }
+}
+
+// set null element in list to NA (NA_LOGICAL) recursively
+void null_to_na(List x) {
+  int size = x.size();
+  for (int i = 0; i < size; i++) {
+    SEXP elem = x[i];
+    if (elem == R_NilValue) {
+      x[i] = LogicalVector::create(NA_LOGICAL);
+    } else if(TYPEOF(elem) == VECSXP) {
+      null_to_na(elem);
+    }
+  }
+}
+
+// tranpose a nested list
+// e.g. list(a=list(x=1,y=2,z=3), b=list(x=1,y=2,z=3)) =>
+//        list(x=list(a=1,b=1), y=list(a=2,b=2), z=list(a=3,b=3))
+// x is the list to tranpose
+// out is a pre-allocated correctly-typed output placeholder
+// type coercion is applied according to `out` type as necessary
+// [[Rcpp::export(name = ".transpose")]]
+List transpose(List x, List out) {
+  int rows = x.size();
+  if(rows == 0) return out;
+
+  int cols = out.size();
+  if(cols == 0) return out;
+
+  for (int c = 0; c < cols; c++) {
+    SEXP column = VECTOR_ELT(out, c);
+    switch(TYPEOF(column)) {
+    case VECSXP: { // list
+      for (int r = 0; r < rows; r++) {
+        SEXP row = VECTOR_ELT(x, r);
+        SEXP cell = VECTOR_ELT(row, c);
+        if (cell != R_NilValue) {
+          if(TYPEOF(cell) == VECSXP) {
+            null_to_na(cell);
```

Comment:
Hmm, note that this line makes the new situation semantically different from the previous pure R code. We now could possibly be modifying the input, which I'm not crazy about. On the other hand, I can see the benefits of in-place modification. In either case, this should be documented in a comment somewhere, probably right above the `transpose` call in `json.tabular.to.data.frame.R`.